performance_test/src/communication_abstractions/opendds_communicator.hpp

```diff
@@ -0,0 +1,344 @@
+  {
+    if (CORBA::is_nil(m_datareader)) {
+      DDS::Subscriber_ptr subscriber = nullptr;
+      DDS::DataReaderQos dr_qos;
+      ResourceManager::get().opendds_subscriber(subscriber, dr_qos);
+
+      dr_qos.resource_limits.max_samples = 32;
+      dr_qos.resource_limits.max_instances = 10;
+      dr_qos.resource_limits.max_samples_per_instance = 32;
+      /* if there are more remote writers, you need to increase these limits */
+      
+      //These are not available in OpenDDS, RTI tricks
+      //dr_qos.reader_resource_limits.max_remote_writers = 10;
+      //dr_qos.reader_resource_limits.max_remote_writers_per_instance = 10;
+
+      OpenDdsQOSAdapter qos_adapter(m_ec.qos());
+      qos_adapter.apply_dr(dr_qos);
+
+      /* Only DDS_DATA_AVAILABLE_STATUS supported currently */
+      m_datareader = subscriber->create_datareader(
+        m_topic,
+        dr_qos,
+        nullptr,
+        OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+
+      if (CORBA::is_nil(m_datareader)) {
+        throw std::runtime_error("datareader == nullptr");
+      }
+
+      m_condition = m_datareader->get_statuscondition();
+      m_condition->set_enabled_statuses(DDS::DATA_AVAILABLE_STATUS);
+      m_waitset.attach_condition(m_condition);
+
+      m_typed_datareader = DataReaderType::_narrow(m_datareader);
+      if (m_typed_datareader == nullptr) {
+        throw std::runtime_error("m_typed_datareader == nullptr");
+      }
+
+      //Again, RTI trick. No wonder
+      //if (!m_condition_seq.ensure_length(2, 2)) {
+      //  throw std::runtime_error("Error ensuring length of active_conditions_seq.");
+      //}
+    }
+
+    if (!m_ec.no_waitset()) {
+      DDS::Duration_t wait_timeout = {15, 0};
+      m_waitset.wait(m_condition_seq, wait_timeout);
+    }
+
+    auto ret = m_typed_datareader->take(m_data_seq, m_sample_info_seq, DDS::LENGTH_UNLIMITED,
+        DDS::ANY_SAMPLE_STATE, DDS::ANY_VIEW_STATE,
+        DDS::ANY_INSTANCE_STATE);
+    if (ret == DDS::RETCODE_OK) {
+      lock();
+      for (decltype(m_data_seq.length()) j = 0; j < m_data_seq.length(); ++j) {
+        const auto & data = m_data_seq[j];
+        if (m_sample_info_seq[j].valid_data) {
+          if (m_prev_timestamp >= data.time_) {
+            throw std::runtime_error(
+                    "Data consistency violated. Received sample with not strictly older timestamp. "
+                    "Time diff: " + std::to_string(
+                      data.time_ - m_prev_timestamp) +
+                    " Data Time: " +
+                    std::to_string(data.time_)
+            );
+          }
+          m_prev_timestamp = data.time_;
+          update_lost_samples_counter(data.id_);
+          add_latency_to_statistics(data.time_);
+          increment_received();
+        }
+      }
+      unlock();
+
+      if (m_ec.roundtrip_mode() == ExperimentConfiguration::RoundTripMode::RELAY) {
+        throw std::runtime_error("Round trip mode is not implemented for Connext DDS Micro!");
```

Comment:
Change the error string to reference OpenDDS

Reply:
done